user/tests/wiring/no_fixture_power_saving/power_saving_mode.cpp

```diff
@@ -0,0 +1,235 @@
+    Particle.function("fnlp1", fnLp1);
+}
+
+test(POWER_SAVING_01_particle_publish_publishes_an_event_after_low_power_active) {
+    if (modem_type != DEV_SARA_R510) {
+        skip();
+        return;
+    }
+
+    bool publishResult = false;
+    bool lowPowerResult = false;
+    int attempts = 0;
+    uint32_t period = 0;
+    const uint32_t WAIT_FOR_LOW_POWER_ACTIVE_MS = 20000;
+    const uint32_t WAIT_FOR_LOW_POWER_MEAS_MS = 10000;
+    const uint32_t LOW_POWER_ATTEMPTS_MAX = 10; // XXX: extreme cases take up to 2.5 minutes to drop into low power mode
+    do {
+        Particle.connect();
+        assertTrue(waitFor(Particle.connected, CLOUD_CONNECT_TIMEOUT));
+        delay(WAIT_FOR_LOW_POWER_ACTIVE_MS); // wait for UPSV=1 default delay of ~9.2s before modem drops into low power mode idle mode.
+
+        period = measureAvgPeriodMs(CTS1, WAIT_FOR_LOW_POWER_MEAS_MS);
+        if (period <= 1280 + 128 && period >= 1280 - 128) { // 1.28s +/- 10% if in low power mode
+            lowPowerResult = true;
+        } else {
+            lowPowerResult = false;
+        }
+
+        publishResult = Particle.publish("my_event_low_power", "event data low power", PRIVATE | WITH_ACK);
+        // Log.info("period: %lu, publishResult: %d", period, publishResult);
+    } while (!(publishResult && lowPowerResult) && attempts++ < LOW_POWER_ATTEMPTS_MAX);
+    assertMoreOrEqual(period, 1280 - 128); // 1.28s +/- 10% if in low power mode
+    assertLessOrEqual(period, 1280 + 128);
+    assertTrue(publishResult);
+}
+
+test(POWER_SAVING_02_register_function_and_connect_to_cloud) {
+    if (modem_type != DEV_SARA_R510) {
+        skip();
+        return;
+    }
+
+    bool lowPowerResult = false;
+    int attempts = 0;
+    uint32_t period = 0;
+    const uint32_t WAIT_FOR_LOW_POWER_ACTIVE_MS = 20000;
+    const uint32_t WAIT_FOR_LOW_POWER_MEAS_MS = 10000;
+    const uint32_t LOW_POWER_ATTEMPTS_MAX = 10; // XXX: extreme cases take up to 2.5 minutes to drop into low power mode
+    do {
+        Particle.connect();
+        assertTrue(waitFor(Particle.connected, CLOUD_CONNECT_TIMEOUT));
+        delay(WAIT_FOR_LOW_POWER_ACTIVE_MS); // wait for UPSV=1 default delay of ~9.2s before modem drops into low power mode idle mode.
+
+        period = measureAvgPeriodMs(CTS1, WAIT_FOR_LOW_POWER_MEAS_MS);
+        if (period <= 1280 + 128 && period >= 1280 - 128) { // 1.28s +/- 10% if in low power mode
+            lowPowerResult = true;
+        } else {
+            lowPowerResult = false;
+        }
+        // Log.info("period: %lu", period);
+
+    } while (!lowPowerResult && attempts++ < LOW_POWER_ATTEMPTS_MAX);
+    assertMoreOrEqual(period, 1280 - 128); // 1.28s +/- 10% if in low power mode
+    assertLessOrEqual(period, 1280 + 128);
+}
+
+test(POWER_SAVING_03_call_function_and_check_return_value_after_low_power_active) {
+    if (modem_type != DEV_SARA_R510) {
+        skip();
+        return;
+    }
+
+    // See no_fixture_long_running.spec.js
```

Comment:
The implication is that we confirmed the device is in power save mode in test 2, so test 3 can continue on the test runner side (which tests that the cloud side can call a function when the device is in low power mode)?

Reply:
Correct.